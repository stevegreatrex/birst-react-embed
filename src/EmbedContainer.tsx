import * as React from 'react';
import { Props } from 'react';

export const EmbedContainer = (props: Props<{}>) => (
	<div
		style={{
			display: 'flex',
			flexDirection: 'column',
			alignContent: 'center',
			justifyContent: 'center',
			fontFamily: 'Segoe UI, Tahoma, Geneva, Verdana, sans-serif',
			overflowX: 'hidden'
		}}>
		{props.children}
	</div>
);
