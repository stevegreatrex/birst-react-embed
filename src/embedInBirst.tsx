import * as React from 'react';
import { ReactChildren } from 'react';
import { render } from 'react-dom';
import { EmbedContainer } from './EmbedContainer';

const attributeName = 'data-birst-content';

export function attachElement<TParams>(
	createEmbeddedElement: (params: TParams) => JSX.Element
) {
	return (element: HTMLElement) => {
		const serializedParameters = element.getAttribute(attributeName);
		const parameters: TParams = serializedParameters
			? JSON.parse(serializedParameters)
			: {};

		render(
			<EmbedContainer>{createEmbeddedElement(parameters)}</EmbedContainer>,
			element
		);
	};
}

export function embedInBirst<TParams>(
	createEmbeddedElement: (params: TParams) => JSX.Element
) {
	Array.from(document.querySelectorAll(`[${attributeName}]`)).forEach(
		attachElement(createEmbeddedElement)
	);
}
