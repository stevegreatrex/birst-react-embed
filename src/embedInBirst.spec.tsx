import * as React from 'react';
import { EmbedContainer } from './EmbedContainer';
import { attachElement } from './embedInBirst';
import * as ReactDom from 'react-dom';

describe('attachElement', () => {
	it('renders into target element', () => {
		const TestComponent = (props: {}) => <div id="content" />;
		const element = document.createElement('div');

		attachElement(() => <TestComponent />)(element);
		expect(element).toMatchSnapshot();
	});

	it('passes parsed props into component', () => {
		const TestComponent = (props: { id?: string }) => <div id={props.id} />;
		const element = document.createElement('div');
		element.setAttribute('data-birst-content', '{ "id": "test" }');

		attachElement(props => <TestComponent {...props} />)(element);
		expect(element.querySelector('div#test')).toBeTruthy();
		expect(element).toMatchSnapshot();
	});
});
