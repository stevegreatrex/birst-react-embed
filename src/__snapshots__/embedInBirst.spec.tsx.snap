// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attachElement passes parsed props into component 1`] = `
<div
  data-birst-content="{ \\"id\\": \\"test\\" }"
>
  <div
    data-reactroot=""
    style="display: flex; font-family: Segoe UI, Tahoma, Geneva, Verdana, sans-serif; overflow-x: hidden;"
  >
    <div
      id="test"
    />
  </div>
</div>
`;

exports[`attachElement renders into target element 1`] = `
<div>
  <div
    data-reactroot=""
    style="display: flex; font-family: Segoe UI, Tahoma, Geneva, Verdana, sans-serif; overflow-x: hidden;"
  >
    <div
      id="content"
    />
  </div>
</div>
`;
